src/pybullet_helpers/ikfast/utils.py

```diff
@@ -0,0 +1,316 @@
+# pylint: skip-file
+
+from __future__ import annotations
+
+import logging
+import random
+import time
+from itertools import chain, islice
+from numbers import Number
+from typing import TYPE_CHECKING, Callable, Generator, List, Sequence, Tuple, \
+    Union
+
+import numpy as np
+
+from predicators.src.pybullet_helpers.geometry import Pose, matrix_from_quat, \
+    multiply_poses
+from predicators.src.pybullet_helpers.ikfast import IKFastInfo
+from predicators.src.pybullet_helpers.ikfast.load import import_ikfast
+from predicators.src.pybullet_helpers.joint import get_joint_infos, \
+    get_joint_lower_limits, get_joint_upper_limits, get_joints, \
+    prune_fixed_joints, violates_joint_limits
+from predicators.src.pybullet_helpers.link import get_link_ancestors, \
+    get_link_from_name, get_link_parent, get_link_pose, \
+    get_parent_joint_from_link, get_relative_link_pose
+from predicators.src.settings import CFG
+from predicators.src.structs import JointsState
+
+"""
+Note: modified from pybullet-planning for our use cases. Minimal changes to the actual
+logic have been made.
+
+https://github.com/caelan/pybullet-planning/blob/master/pybullet_tools/ikfast/ikfast.py
+"""
+
```

Comment:
I see what you mean about this IKFast stuff being tricky. my hunch is that Caelan accumulated a lot of bells and whistles to his implementation that he needed for different purposes, and we might not need all of them for our purposes. do you think that is right? if so, is there a way that we could simplify the code here and revisit it later as needed?

Reply:
Yep - I think some of the bells and whistles we do need - e.g. returning solution with minimum distance from current configuration but other stuff like the generators we could simplify a bit. 

So all the methods from `def get_length` to `def uniform_generator(d):` could be simplified.